### ISSUE TITLE
confint for random effects in more complex cases

### ISSUE THREAD

**Issue by strengejacke** [glmmTMB/glmmTMB#817]
I'd like to calculate confidence intervals for the random effect variances. This works in "simple" cases, but `confint()` fails for more complex RE structures (might also be due the warnings during model fitting, I don't know). Do you think these issues can be solved?

``` r
library(glmmTMB)
library(lme4)
#> Loading required package: Matrix

data(sleepstudy)
data(cake)
set.seed(123)
sleepstudy$Months <- sample(1:4, nrow(sleepstudy), TRUE)

m1 <- glmmTMB(Reaction ~ Days + (Days | Subject), data = sleepstudy)
m2 <- glmmTMB(angle ~ temperature + (temperature | recipe) + (temperature | replicate), data = cake)
#> Warning in fitTMB(TMBStruc): Model convergence problem; non-positive-definite
#> Hessian matrix. See vignette('troubleshooting')
#> Warning in fitTMB(TMBStruc): Model convergence problem; iteration limit reached
#> without convergence (10). See vignette('troubleshooting')
m3 <- glmmTMB(angle ~ temperature + (temperature | recipe), data = cake)
#> Warning in fitTMB(TMBStruc): Model convergence problem; non-positive-definite
#> Hessian matrix. See vignette('troubleshooting')
m4 <- glmmTMB(angle ~ temperature + (temperature | replicate), data = cake)
#> Warning in fitTMB(TMBStruc): Model convergence problem; non-positive-definite
#> Hessian matrix. See vignette('troubleshooting')
#> Warning in fitTMB(TMBStruc): Model convergence problem; false convergence (8).
#> See vignette('troubleshooting')
m5 <- glmmTMB(Reaction ~ Days + (Days + Months | Subject), data = sleepstudy)
#> Warning in fitTMB(TMBStruc): Model convergence problem; non-positive-definite
#> Hessian matrix. See vignette('troubleshooting')

confint(m1, parm = "theta_", method = "wald", level = .9)
#>                             5 %       95 %    Estimate
#> Std.Dev.(Intercept)  16.1689591 34.9753951 23.78057467
#> Std.Dev.Days          4.0627802  8.0442555  5.71682093
#> Cor.Days.(Intercept) -0.4147801  0.5263468  0.08132635

confint(m2, parm = "theta_", method = "wald", level = .9)
#> Warning in sqrt(diag(vv)): NaNs produced
#> Warning in sqrt(diag(object$cov.fixed)): NaNs produced
#> Warning in sqrt(as.numeric(object$diag.cov.random)): NaNs produced
#> Warning in sqrt(diag(object$cov.fixed)): NaNs produced
#> Error in dimnames(x) <- dn: length of 'dimnames' [2] not equal to array extent

confint(m3, parm = "theta_", method = "wald", level = .9)
#> Warning in sqrt(diag(vv)): NaNs produced
#> Warning in sqrt(diag(vv)): NaNs produced
#> Warning in sqrt(as.numeric(object$diag.cov.random)): NaNs produced
#> Warning in sqrt(diag(object$cov.fixed)): NaNs produced
#> Warning in sqrt(as.numeric(object$diag.cov.random)): NaNs produced
#> Error in if (all(substr(rownames(ci), 1, 5) == "cond.")) {: missing value where TRUE/FALSE needed

confint(m4, parm = "theta_", method = "wald", level = .9)
#> Warning in sqrt(diag(vv)): NaNs produced
#> Warning in sqrt(diag(object$cov.fixed)): NaNs produced
#> Warning in sqrt(diag(object$cov.fixed)): NaNs produced
#> Error in if (all(substr(rownames(ci), 1, 5) == "cond.")) {: missing value where TRUE/FALSE needed

confint(m5, parm = "theta_", method = "wald", level = .9)
#> Warning in sqrt(diag(vv)): NaNs produced
#> Warning in sqrt(diag(vv)): NaNs produced
#> Warning in sqrt(as.numeric(object$diag.cov.random)): NaNs produced
#> Warning in sqrt(diag(object$cov.fixed)): NaNs produced
#> Warning in sqrt(as.numeric(object$diag.cov.random)): NaNs produced
#> Error in if (all(substr(rownames(ci), 1, 5) == "cond.")) {: missing value where TRUE/FALSE needed
```

<sup>Created on 2022-05-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by bbolker** [comment]
In any case glmmTMB is much less robust for singular models than lme4 is: (1) lme4 stops at the singular boundary, glmmTMB fits on a log-sd scale so tries to move the estimate to `-Inf`; (2) glmmTMB relies much more heavily on Wald estimates (we would essentially never expect Wald estimates to work for singular models - either they're at a boundary in which case Wald estimates aren't applicable, or they're 'approaching infinity' (if we're fitting on the log-sd scale), so the surface is very flat and the Wald estimates break down ...

But (a) we should get useful warnings or errors in any case; (b) these might be good test cases for `diagnose()`

**Comment by strengejacke** [comment]
FWIW, all models fit with lme4 work fine, we use merDeriv to get the vcov for random effects and thus are able to compute confidence intervals... (See https://github.com/easystats/parameters/blob/main/tests/testthat/test-random_effects_ci.R, where all tests pass...)

The (somewhat ugly) code to get CIs is [here](https://github.com/easystats/parameters/blob/6b8c2b9b3ceff831aa4c4ef63b7d9e8a00d001fa/R/extract_random_variances.R#L302).

**Comment by bbolker** [comment]
OK.  Do you know if the confidence intervals that you get in this case actually **make any sense**?  I would expect the Wald CIs to be really, really bad in this case (this all goes back to the reason that Doug Bates didn't report standard errors of  RE parameters for `lme4` in the first place ... Wald CIs on the log scale are better behaved than on the linear scale, but they're still not all that reliable, and Wald CIs should *never* work [or only work by accident] when the fit is singular ...). My strong advice would be to return `NA` for Wald CIs when it looks like the model is singular ...

(Nevertheless, this is still an interesting thing to dig into)

**Comment by strengejacke** [comment]
Let's take the simpler example from those 5 models (the others do have rather long outputs). You can see that for the correlations, the CIs span over the whole possible range (-1 to 1), and literally the CIs for the SD of "Month" also spans over the complete range of possible values. Thus, "they make sense" in the meaning that they reflect the "uncertainty" quite well, but the CIs are of course pretty "useless" regarding any kind of precision.

Maybe we should yield a warning when the model is a singular fit, that SEs and CIs for random effects are not meaningful.

``` r
library(lme4)
#> Loading required package: Matrix
library(parameters)
data(sleepstudy)
set.seed(123)
sleepstudy$Months <- sample(1:4, nrow(sleepstudy), TRUE)
m5 <- lmer(Reaction ~ Days + (Days + Months | Subject), data = sleepstudy)
#> boundary (singular) fit: see help('isSingular')
model_parameters(m5)
#> # Fixed Effects
#> 
#> Parameter   | Coefficient |   SE |           95% CI | t(171) |      p
#> ---------------------------------------------------------------------
#> (Intercept) |      251.43 | 6.80 | [238.00, 264.85] |  36.96 | < .001
#> Days        |       10.45 | 1.55 | [  7.40,  13.50] |   6.76 | < .001
#> 
#> # Random Effects
#> 
#> Parameter                       | Coefficient |    SE |            95% CI
#> -------------------------------------------------------------------------
#> SD (Intercept: Subject)         |       25.11 |  8.80 | [12.64,    49.89]
#> SD (Days: Subject)              |        5.92 |  6.98 | [ 0.59,    59.76]
#> SD (Months: Subject)            |        0.32 | 36.85 | [ 0.00, 4.23e+98]
#> Cor (Intercept~Days: Subject)   |        0.09 |  0.39 | [-0.59,     0.70]
#> Cor (Intercept~Months: Subject) |       -0.61 |  1.86 | [-1.00,     1.00]
#> Cor (Days~Months: Subject)      |       -0.85 |  6.96 | [-1.00,     1.00]
#> SD (Residual)                   |       25.60 |  1.60 | [22.65,    28.93]
#> 
#> Uncertainty intervals (equal-tailed) and p-values (two-tailed) computed
#>   using a Wald t-distribution approximation.
```

<sup>Created on 2022-05-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
@bwiernik @mattansb what's your opinion? Return `NA`, or keep the CI bounds at their extreme value, but yield a warning about the possible "useless" range the CIs cover?

**Comment by bwiernik** [comment]
I think NA for Wald CIs when the fit is singular is a good idea

**Comment by bbolker** [comment]
Only issue is that we don't have a definition of "singular" for glmmTMB (yet) - it will have to be "log-sd is small enough *or* estimated correlation matrix is close enough to non-pos-definite" (or something)

**Comment by bwiernik** [comment]
What's the estimation metric for correlations? atanh?

**Comment by mattansb** [comment]
> I think NA for Wald CIs when the fit is singular is a good idea

Agree

**Comment by bbolker** [comment]
x/(1+x^2). From [the covstruct vignette](https://glmmtmb.github.io/glmmTMB/articles/covstruct.html):

> For an unstructured matrix of size n, parameters 1:n represent the log-standard deviations while the remaining n(n-1)/2 (i.e. (n+1):(n:(n*(n+1)/2))) are the elements of the scaled Cholesky factor of the correlation matrix, filled in row-wise order (see [TMB documentation](http://kaskr.github.io/adcomp/classUNSTRUCTURED__CORR__t.html)). In particular, if L is the lower-triangular matrix with 1 on the diagonal and the correlation parameters in the lower triangle, then the correlation matrix is defined as Σ=D^(−1/2) L L^⊤D^(−1/2), where D=diag(LL^⊤). For a single correlation parameter θ_0, this works out to ρ=θ_0/(1+θ_0^2).

(Now that I look at it the last sentence seems wrong/implausible, I will have to look more carefully!)

**Comment by bwiernik** [comment]
Okay, Choleskey factors, that makes sense. Will think a bit about heuristics to detect singularities

**Comment by strengejacke** [comment]
I can confirm that it works now (i.e. no errors, returns `NaN` for some values). But you have changes the row names, is that intentional?

was: Std.Dev.(Intercept)
is: Std.Dev.(Intercept)|Subject

If this change will be kept, I need to adjust `model_parameters()` accordingly to find the correct CIs (based on row name matching).

**Comment by bbolker** [comment]
I do think I want to change the names; they didn't have the grouping variable in them before, which makes them potentially non-unique!  I do realize this is a breaking change, we'll definitely have to check downstream packages/note this in `NEWS` ...

**Comment by bwiernik** [comment]
I like the new names much better.

**Comment by strengejacke** [comment]
I think this issue is closed by #824?

**Comment by bbolker** [comment]
Yes; I should open a new issue for dealing with mapped parameters (WIP documented [here])
